pennylane/ops/op_math/condition.py

```diff
@@ -647,39 +665,134 @@ def wrapper(*args, **kwargs):
+        """
+        if not jax.config.jax_dynamic_shapes:
+            # fallback to normal behavior
+            return jaxpr_trace.default_process_primitive(cond_prim, tracers, params)
+        true_outvars = params["jaxpr_branches"][0].outvars
+
+        env = {}  # branch var to new equation var
+        out_tracers, returned_vars = tuple(
+            zip(*(_tracer_and_outvar(jaxpr_trace, var, env) for var in true_outvars))
+        )
+
+        invars = [jaxpr_trace.getvar(x) for x in tracers]
+        eqn = pe.new_jaxpr_eqn(
+            invars,
+            returned_vars,
+            cond_prim,
+            params,
+            jax.core.no_effects,
+        )
+        jaxpr_trace.frame.add_eqn(eqn)
+        return out_tracers
+
+    pe.custom_staging_rules[cond_prim] = custom_staging_rule
+
+
+def _shape_error(branch_type, branch_index, i, outval, expected_outval):
+    raise ValueError(
+        f"Mismatch in output abstract values in {branch_type} branch "
+        f"#{branch_index} at position {i}: "
+        f"{outval} vs {expected_outval}."
+    )
+
+
 def _validate_abstract_values(
-    outvals: list, expected_outvals: list, branch_type: str, index: int = None
+    outvals: list, expected_outvals: list, branch_type: str, branch_index: int
 ) -> None:
     """Ensure the collected abstract values match the expected ones."""
+    import jax  # pylint: disable=import-outside-toplevel
 
     if len(outvals) != len(expected_outvals):
-        raise ValueError(
+        msg = (
             f"Mismatch in number of output variables in {branch_type} branch"
-            f"{'' if index is None else ' #' + str(index)}: "
-            f"{len(outvals)} vs {len(expected_outvals)}"
+            f" #{branch_index}: {len(outvals)} vs {len(expected_outvals)} "
+            f" for {outvals} and {expected_outvals}"
         )
+        if jax.config.jax_dynamic_shapes:
+            msg += "\n This may be due to different sized shapes when dynamic shapes are enabled."
+        raise ValueError(msg)
 
     for i, (outval, expected_outval) in enumerate(zip(outvals, expected_outvals)):
-        if outval != expected_outval:
-            raise ValueError(
-                f"Mismatch in output abstract values in {branch_type} branch"
-                f"{'' if index is None else ' #' + str(index)} at position {i}: "
-                f"{outval} vs {expected_outval}"
-            )
+        if jax.config.jax_dynamic_shapes:
+            # we need to be a bit more manual with the comparison.
+            if type(outval) != type(expected_outval):  # pylint: disable=unidiomatic-typecheck
+                _shape_error(branch_type, branch_index, i, outval, expected_outval)
+            if getattr(outval, "dtype", None) != getattr(expected_outval, "dtype", None):
+                _shape_error(branch_type, branch_index, i, outval, expected_outval)
+
+            for s1, s2 in zip(getattr(outval, "shape", ()), getattr(expected_outval, "shape", ())):

```

Comment:
```suggestion
            for s1, s2 in zip(getattr(outval, "shape", ()), getattr(expected_outval, "shape", ()), strict=True):

```

Reply:
updating locally to clean the line up at the same time.